### PR TITLE
Fix writing of Nullable attributes

### PR DESCRIPTION
**Pull request by marcelveldt**
Some Matter Attributes require sending a Null value to reset it to its device/factory default.
This was not working correctly, especially in the case if the annotation was a union (e/.g. an optional int).

This issue is now fixed so the write_attribute can correctly understand that it actually needs to write the sdk's Null if python sends a None.

Also added a quick test case

### CODE REVIEW

File: tests/common/test_parser.py [hunk 1]
```diff
@@ -110,3 +111,22 @@ def test_dataclass_from_dict():
     # test NOCStruct.noc edge case
     res = parse_value("NOCStruct.noc", 5, bytes)
     assert res == b""
+
+
+def test_parse_value():
+    """Test special cases in the parse_value helper."""
+    # ruff: noqa: PT011
```
Comment (agners) [inline tests/common/test_parser.py:118]: I think this is not necessary anymore

File: matter_server/common/helpers/util.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def parse_value(
     value: Any,
     value_type: Any,
     default: Any = MISSING,
-    allow_none: bool = True,
+    allow_none: bool = False,
```
Comment (agners) [inline matter_server/common/helpers/util.py:110]: Seems that with this PR `allow_none` is not really used anymore? Can we drop it completely?

Comment (marcelveldt) [reply matter_server/common/helpers/util.py:110]: II couldnt spot where it was used but was scared to remove it

Comment (marcelveldt) [reply matter_server/common/helpers/util.py:110]: it is still in use (within dataclass_from_dict)

File: tests/common/test_parser.py (hunk 1 above)
Comment (marcelveldt) [reply tests/common/test_parser.py]: cleaned up

File: matter_server/common/helpers/util.py (hunk 2 above)
Comment (agners) [reply matter_server/common/helpers/util.py:110]: Hm, I see we pass it from `dataclass_from_dict()` to `parse_value()`. But since `dataclass_from_dict()` is in our code too, we can just drop it there no? It doesn't serve any purpose and might only lead to confusion. :man_shrugging:

Comment (marcelveldt) [reply matter_server/common/helpers/util.py:110]: No, dataclass_from_dict is used in HA we cant just drop that - it does serve a purpose, parse it strict or not strict.

In strict mode a None value is seen as error if that is not in the annotations.

Comment (agners) [reply matter_server/common/helpers/util.py:110]: Sorry, I didn't meant to suggest to drop `dataclass_from_dict`. I am suggesting to drop `allow_none` from parse_value()`. It needs a `dataclass_from_dict` update, but it should be a simple oneline change:


```patch
--- a/matter_server/common/helpers/util.py
+++ b/matter_server/common/helpers/util.py
@@ -272,7 +272,6 @@ def dataclass_from_dict(
                 dict_obj.get(field.name),
                 type_hints[field.name],
                 field.default,
-                allow_none=not strict,
                 allow_sdk_types=allow_sdk_types,
             )
             for field in dc_fields
```

Comment (marcelveldt) [reply matter_server/common/helpers/util.py:110]: No, because that is exactly where we use it!
if the dataclass_from_dict is strict we want it to raise on None values

**Review by agners — APPROVED** [review]

File: matter_server/common/helpers/util.py (hunk 2 above)
Comment (marcelveldt) [reply matter_server/common/helpers/util.py:110]: Almost weekend ;-)
